sound/soc/sof/sof-audio.c

```diff
@@ -11,6 +11,561 @@
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_source_path(widget, p)
+				if (is_sof_widget(p->source)) {
+					ret = sof_route_setup(sdev, p->source, widget);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	}
+
+	return 0;
+}
+
+static int sof_widget_list_dai_config_setup(struct snd_sof_dev *sdev,
+					    struct snd_soc_dapm_widget_list *list)
+{
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_widget *swidget;
+	struct snd_sof_dai *dai;
+	int ret, i;
+
+	for_each_dapm_widgets(list, i, widget) {
+		swidget = widget->dobj.private;
+		if (!swidget)
+			continue;
+
+		if  (swidget->id == snd_soc_dapm_dai_in || swidget->id == snd_soc_dapm_dai_out) {
+			dai = swidget->private;
+
+			ret = sof_dai_config_setup(sdev, dai);
+			if (ret < 0) {
+				dev_err(sdev->dev, "error: failed to load dai config for DAI %s\n",
+					swidget->widget->name);
+				return ret;
+			}
+		}
+	}
+
+	return 0;
+}
+
+int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, int dir)
+{
+	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
+	struct sof_ipc_fw_version *v = &sdev->fw_ready.version;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_widget *w, *_w;
+	int pipe_widget_count = 0;
+	int i, ret;
+	int j;
+	/* nothing to set up */
+	if (!list)
+		return 0;
+
+	/*
+	 * Dynamic pipeline feature is only available from 3.19 onwards. So for older firmware,
+	 * we don't need to set up the widgets but we do need to send the DAI_CONFIG IPC for DAI
+	 * widgets.
+	 */
+	if (v->abi_version < SOF_ABI_VER(3, 19, 0))
+		return sof_widget_list_dai_config_setup(sdev, list);
+
+	/* set up widgets in the list */
+	for_each_dapm_widgets(list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+		struct snd_sof_widget *pipe_widget, *temp_pipe_widget;
+		bool pipeline_found = false;
+
+		if (!swidget)
+			continue;
+
+		ret = sof_widget_setup(sdev, swidget);
+		if (ret < 0)
+			goto widget_free;
+
+		/* find pipeline widget for the pipeline that this widget belongs to */
+		pipe_widget = snd_sof_pipeline_find(sdev, swidget->pipeline_id);
+		if (!pipe_widget) {
+			ret = -EINVAL;
+			goto widget_free;
+		}
+
+		/* Add it to the spcm pipeline list if it isn't part of it already. */
+		list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+			if (w->comp_id == pipe_widget->comp_id) {
+				pipeline_found = true;
+				break;
+			}
+		}
+		if (!pipeline_found) {
+			/*
+			 * Pipeline widgets can only be set up after the scheduling
+			 * widget has been set up. Create a temp widget to save it to the spcm
+			 * pipeline list as the original widget is already part of the core's
+			 * widget list.
+			 */
+			temp_pipe_widget = kmemdup(pipe_widget, sizeof(*pipe_widget), GFP_KERNEL);
+			if (!temp_pipe_widget) {
+				ret = -ENOMEM;
+				goto widget_free;
+			}
+
+			list_add(&temp_pipe_widget->list, &spcm->stream[dir].pipeline_list);
+		}
```

Comment:
@ranj063 I get your idea about pipeline list usage.  pipeline widget is only added to pipeline list not widget list in sdev in topology.c. I want to add  struct list_head pipeline_list in widget to link pipeline widget to spcm.pipeline_list so we don't need to duplicate another pipeline widget  and also avoid searching original pipeline widget in following part such as free pipeline setup pipeline. And also we can link other widgets to its correspoding pipeline widget, so that we can quickly find the widgets of a pipeline.  Now for ipc4, we need to first check each widget to get total memory usage, then create pipeline and widgets later, so I need to speed it up to avoid searching each other

Reply:
I just find your explanation above. I will consider my algorithm again. Recently I checked windows driver again and found my implementation is very simple so I am enhancing it now.